cupy/linalg/_einsum.py

```diff
@@ -150,6 +150,9 @@ def _parse_einsum_input(args):
                 msg + ' operands provided to einstein sum function than '
                 'specified in the subscripts string')
 
+        # NumPy ignores 'weak' scalars and always returns i64/f64
+        operands = [cupy.asarray(op) for op in operands]
```

Comment:
Maybe should change that one day, but not today.  `i64` I assume is true, `f64` I would think not if inputs are float32?

I suppose since scalars are an odd input to einsum, not changing it should be OK or maybe we should fix this one in NumPy?

Reply:
Indeed, `s/always/sometimes`:

```
In [5]: np.einsum('i->', np.arange(8, dtype=np.float32)).dtype
Out[5]: dtype('float32')

In [6]: np.einsum(', i->', 3, np.arange(8, dtype=np.float32)).dtype
Out[6]: dtype('float64')

In [7]: np.__version__
Out[7]: '2.0.0rc2'
```

numpy bug?

Reply:
So this patch correctly avoids using weak promotion in einsum.